odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2248 @@
+        4.0
+        """
+        return self.weighting.inner(x1, x2)
+
+    def _multiply(self, x1, x2, out):
+        """Entry-wise product of two tensors, assigned to out.
+
+        Parameters
+        ----------
+        x1, x2 : `CupyTensor`
+            Factors in the product.
+        out : `CupyTensor`
+            Tensor to which the result is written.
+
+        Examples
+        --------
+        Out-of-place evaluation:
+
+        >>> r3 = odl.rn(3, impl='cupy')
+        >>> x = r3.element([1, 2, 3])
+        >>> y = r3.element([-1, 0, 1])
+        >>> r3.multiply(x, y)
+        rn(3, impl='cupy').element([-1.,  0.,  3.])
+
+        In-place:
+
+        >>> out = r3.element()
+        >>> result = r3.multiply(x, y, out=out)
+        >>> result
+        rn(3, impl='cupy').element([-1.,  0.,  3.])
+        >>> result is out
+        True
+        """
+        x1.ufuncs.multiply(x2, out=out)
+
+    def _divide(self, x1, x2, out):
+        """Entry-wise division of two tensors, assigned to out.
+
+        Parameters
+        ----------
+        x1, x2 : `CupyTensor`
+            Dividend and divisor in the quotient.
+        out : `CupyTensor`
+            Tensor to which the result is written.
+
+        Examples
+        --------
+        Out-of-place evaluation:
+
+        >>> r3 = odl.rn(3, impl='cupy')
+        >>> x = r3.element([1, 2, 3])
+        >>> y = r3.element([-1, 2, 1])
+        >>> r3.divide(x, y)
+        rn(3, impl='cupy').element([-1.,  1.,  3.])
+
+        In-place:
+
+        >>> out = r3.element()
+        >>> result = r3.divide(x, y, out=out)
+        >>> result
+        rn(3, impl='cupy').element([-1.,  1.,  3.])
+        >>> result is out
+        True
+        """
+        x1.ufuncs.divide(x2, out=out)
+
+    def __repr__(self):
+        """Return ``repr(self)``."""
+        if self.ndim == 1:
+            posargs = [self.size]
+        else:
+            posargs = [self.shape]
+
+        if self.is_real:
+            constructor_name = 'rn'
+        elif self.is_complex:
+            constructor_name = 'cn'
+        else:
+            constructor_name = 'tensor_space'
+
+        if (constructor_name == 'tensor_space' or
+                (not self.is_real and not self.is_complex) or
+                self.dtype != self.default_dtype(self.field)):
+            posargs.append(dtype_str(self.dtype))
+
+        optargs = [('impl', self.impl, 'numpy'),  # for the helper functions
+                   ('device', self.device, cupy.cuda.get_device_id())]
+        inner_str = signature_string(posargs, optargs)
+        weight_str = self.weighting.repr_part
+        if weight_str:
+            inner_str += ', ' + weight_str
+
+        return '{}({})'.format(constructor_name, inner_str)
+
+    @property
+    def element_type(self):
+        """`CupyTensor`"""
+        return CupyTensor
+
+    @staticmethod
+    def available_dtypes():
+        """Return the data types available for this space."""
```

Comment:
But in lincomb we only support a very small subset of these?

Reply:
`lincomb` should also support all the others, just many of them without cuBLAS acceleration.

Reply:
BTW, I did some profiling lately, and this function turned out to be a bottleneck. In some PR (I think #1238) I changed this to use a `set` instead of a tuple, which cut down the time to create a new space by quite a bit.
If you chose to do so here as well, be aware that `set` compares by hashing, not by `__eq__`, so you have to use `np.dtype(dtype)` as comparison, otherwise `float` won't be considered valid, for instance.